Gpu levelzero sidecar

tkatila opened the pull request:
Add GPU Levelzero sidecar to allow fetching health data for the GPU devices from the Levelzero API.

As a bonus, this also adds support for detecting Intel GPUs within Windows Subsystem for Linux (WSL).

eero-t reviewed (commented):
Few doc comments.

PS. I would also suggest running following (untested) command on the code to fix PCI abbreviation use in error messages:
`sed -i 's/ pci / PCI /' $(git grep -l ' pci ')`

tkatila commented:
@eero-t  & @uniemimu can you please review?

This probably requires some restructuring after the GPU CDI support has been merged (assuming it will happen first).

(Failed check is due to internal certificate issue, I do not suspect it to be a real problem.)

eero-t commented:
> @eero-t & @uniemimu can you please review?

What are the differences to earlier version?   (you did force push, so github does not show those.)

tkatila commented:
> > @eero-t & @uniemimu can you please review?
> 
> What are the differences to earlier version? (you did force push, so github does not show those.)

https://github.com/intel/intel-device-plugins-for-kubernetes/compare/82a387760e4c2a9a12cc366405e19153f67e7efd..76b1e5a04abc600ec1992a1adf0474d42ffad1b9

~mid third is my changes. Top third and bottom third are baseline changes. But to summarize:
1) Some logging to C code
2) Added temperature read from device and its handling in the plugin

eero-t commented on cmd/gpu_levelzero/zes.c:
```diff
@@ -0,0 +1,441 @@
+        return 0;
+    }
+
+    if (verbosity_level >= verbosity_debug) {
+        fprintf(stderr, "Retrieve memory size for %s\n", bdf_address);
+    }
```
Uh, this check repeated so many times that IMHO there should be helper(s) for them.  For example:
```
static void log(log_level_t level, const char *msg) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s\n", msg);
    }
}

static void log_number(log_level_t level, const char *msg, const char *value) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s: %s\n", msg, value);
    }
}

static void log_string(log_level_t level, const char *msg, const char *value) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s: %s\n", msg, value);
    }
}
```
```suggestion
    log_string(LOG_DEBUG, "Retrieve memory size for", bdf_address);
```

Or if you'd rather have single log function, switch to using `va_args`.

eero-t commented on cmd/gpu_levelzero/zes.c:
```diff
@@ -0,0 +1,441 @@
+
+    // Levelzero does not provide memory details for integrated
+    if (is_integrated(handle)) {
+        if (verbosity_level >= verbosity_debug) {
+            fprintf(stderr, "Device is integrated => no memory\n");
+        }
```
```suggestion
        log(LOG_DEBUG, "Device is integrated => no memory");
```

eero-t commented on cmd/gpu_levelzero/zes.c:
```diff
@@ -0,0 +1,441 @@
+        }
+    }
+
+    if (verbosity_level >= verbosity_debug) {
+        fprintf(stderr, "> Memory size: %ld\n", memory_size);
+    }
```
```suggestion
    log_number(LOG_DEBUG, "> Memory size", memory_size);
```
etc.

eero-t commented on cmd/gpu_levelzero/zes.c:
```diff
@@ -0,0 +1,441 @@
+
+static bool device_enumerated = false;
+
+// error = 1, warning = 2, info = 3
+static const int verbosity_debug = 4;
+static int verbosity_level = 1;
```
```suggestion
typedef enum {
    LOG_ERROR,
    LOG_WARNING,
    LOG_INFO,
    LOG_DEBUG
} log_level_t;

static int verbosity_level = LOG_ERROR;
```

tkatila commented on cmd/gpu_plugin/levelzeroservice/levelzero_service.go:
```diff
@@ -0,0 +1,179 @@
+	GetDeviceMemoryAmount(bdfAddress string) (uint64, error)
+}
+
+type DeviceHealth struct {
+	Memory            bool
+	Bus               bool
+	SoC               bool
+	GlobalTemperature float64
+	GPUTemperature    float64
+	MemoryTemperature float64
+}
```
I split them to different functions, one for health and one for temperatures.

tkatila commented on cmd/gpu_plugin/gpu_plugin.go:
```diff
@@ -575,8 +704,11 @@ func main() {
 	flag.BoolVar(&opts.resourceManagement, "resource-manager", false, "fractional GPU resource management")
+	flag.BoolVar(&opts.wslScan, "wsl", false, "run plugin in WSL environment")
 	flag.IntVar(&opts.sharedDevNum, "shared-dev-num", 1, "number of containers sharing the same GPU device")
+	flag.IntVar(&opts.unhealthyTemperature, "unhealthy-temp", 100, "unhealthy temperature limit")
```
yep, sounds better.

tkatila commented:
Rebased on top of the CDI changes. Also added an e2e test for the levelzero deployment.

eero-t commented on cmd/gpu_levelzero/zes.c:
```diff
@@ -0,0 +1,436 @@
+    return handle;
+}
+
+bool is_integrated(zes_device_handle_t handle)
```
Ditto:
```suggestion
static bool is_integrated(zes_device_handle_t handle)
```

mythi commented:
should it be mentioned somewhere that these new features are not available for the operator users?

tkatila commented:
> should it be mentioned somewhere that these new features are not available for the operator users?

Sure, I'll add a note. This PR is large enough as it is so I didn't want to touch the operator use case. Once this is merged, I'll tackle the operator support.

tkatila commented:
Rebased the content. Added C fixes, fixed compile flags for the build and decided to change the levelzero enabling to a boolean (instead of a unis socket path).

eero-t commented on cmd/gpu_plugin/gpu_plugin.go:
```diff
@@ -651,6 +794,34 @@ func main() {
+			os.Exit(1)
+		}
+
+		if plugin.options.levelzero {
+			klog.Error("WSL requires Level-Zero service to operate. See README for help.")
```
Shouldn't this be rather:
```suggestion
		if !plugin.options.levelzero {
			klog.Error("WSL requires Level-Zero service to operate. See README for help.")
```

But I wonder why L0 use is not enabled automatically if WSL needs it?

tkatila replied:
Yes, this was badly converted from the previous "enable levelzero with socket path" method. And yes, if WSL is enabled, levelzero should be enabled at the same time.

eero-t commented on cmd/gpu_plugin/gpu_plugin.go:
```diff
@@ -627,7 +767,10 @@ func main() {
 	flag.StringVar(&prefix, "prefix", "", "Prefix for devfs & sysfs paths")
 	flag.BoolVar(&opts.enableMonitoring, "enable-monitoring", false, "whether to enable '*_monitoring' (= all GPUs) resource")
 	flag.BoolVar(&opts.resourceManagement, "resource-manager", false, "fractional GPU resource management")
+	flag.BoolVar(&opts.levelzero, "levelzero", false, "Enable retrieval of data from Level-ero interface")
```
```suggestion
	flag.BoolVar(&opts.levelzero, "levelzero", false, "Enable retrieval of data from Level-Zero interface")
```

But I think rather than having this option, L0 should be enabled automatically if features needing it are enabled (WSL, temp limit, or GPU error checks).   And if L0 is not available when those features are requested, plugin should either fail, or log a warning.

tkatila replied:
That is a good point. I'll rewrite this so that each feature has a enable flag.

eero-t commented on cmd/gpu_plugin/gpu_plugin.go:
```diff
@@ -627,7 +767,10 @@ func main() {
 	flag.BoolVar(&opts.enableMonitoring, "enable-monitoring", false, "whether to enable '*_monitoring' (= all GPUs) resource")
 	flag.BoolVar(&opts.resourceManagement, "resource-manager", false, "fractional GPU resource management")
+	flag.BoolVar(&opts.levelzero, "levelzero", false, "Enable retrieval of data from Level-ero interface")
+	flag.BoolVar(&opts.wslScan, "wsl", false, "run plugin in WSL environment")
```
I do not think plugin can invoke WSL VM, but it could be told that it's running in one, and should use WSL devices rather than native ones:
```suggestion
	flag.BoolVar(&opts.wslScan, "wsl", false, "Scan for / use WSL devices")
```

eero-t commented on cmd/gpu_plugin/README.md:
```diff
@@ -56,6 +56,8 @@ For workloads on different KMDs, see [KMD and UMD](#kmd-and-umd).
 |:---- |:-------- |:------- |:------- |
 | -enable-monitoring | - | disabled | Enable '*_monitoring' resource that provides access to all Intel GPU devices on the node, [see use](./monitoring.md) |
 | -resource-manager | - | disabled | Enable fractional resource management, [see use](./fractional.md) |
+| -health-management | - | disabled | Enable health management by requesting data from oneAPI/Level-Zero interface. Requires [GPU Level-Zero](../gpu_levelzero/) sidecar. |
+| -wsl | - | disabled | Adapt plugin to run in the WSL environment. Requires [GPU Level-Zero](../gpu_levelzero/) sidecar. |
```
Potential alternative descriptions:
```suggestion
| -health-management | - | disabled | Disable GPU devices when metrics indicate them not being healthy.  Requires [GPU Level-Zero](../gpu_levelzero/) sidecar. |
| -wsl | - | disabled | Use / scan for WSL environment pass-through GPU devices. Requires [GPU Level-Zero](../gpu_levelzero/) sidecar. |
```

(If descriptions changed here, would need to be changed also elsewhere.)

tkatila replied:
I kept them as they are. But I did add a "health management" section with some explanations.

eero-t reviewed (approved):
Looks OK on quick review.

tkatila commented:
Since I broke wsl in the previous commits, I verified it to work on the current HEAD.

uniemimu commented on cmd/gpu_plugin/gpu_plugin.go:
```diff
@@ -325,15 +337,79 @@ func newDevicePlugin(sysfsDir, devfsDir string, options cliOptions) *devicePlugi
+	if err != nil {
+		klog.Warningf("Device temperature retrieval failed: %v", err)
+
+		return pluginapi.Healthy
```
You might have healthOk = false, yet here you return Healthy. Are you sure you want that?

tkatila replied:
You are correct. In that case the device should be Unhealthy. Actually, if the health indicators say Unhealthy, then it's really no point in reading the temperatures.

mythi reviewed (approved):
lgtm

tkatila commented:
Squashed commits, no code changes.
